File: src/main/java/moe/giga/discord/SettingsManager.java
package moe.giga.discord;

import com.alibaba.fastjson.JSON;
import org.pmw.tinylog.Logger;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Collectors;

public final class SettingsManager {
    private static SettingsManager instance;
    private Settings settings;
    private final Path configFile = new File(".").toPath().resolve("config.json");

    public static SettingsManager getInstance() {
        if (instance == null) {
            instance = new SettingsManager();
        }
        return instance;
    }

    private SettingsManager() {
        if (!configFile.toFile().exists()) {
            Logger.info("Creating Default Settings...");
            Logger.info("You will need to edit config.json with your login information.");
            this.settings = getDefaultSettings();
            saveSettings();
            System.exit(LucoaBot.NEWLY_CREATED_CONFIG);
        }
        loadSettings();
    }

    private Settings getDefaultSettings() {
        Settings newSettings = new Settings();

        newSettings.setBotToken("");

        return newSettings;
    }

    public void saveSettings() {
        try {
            BufferedWriter writer = Files.newBufferedWriter(configFile, StandardCharsets.UTF_8);
            JSON.writeJSONString(writer, this.settings);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void loadSettings() {
        try {
            BufferedReader reader = Files.newBufferedReader(configFile, StandardCharsets.UTF_8);
            String json = reader.lines().collect(Collectors.joining("\n"));
            this.settings = JSON.parseObject(json, Settings.class);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public Settings getSettings() {
        return settings;
    }
}

File: src/main/kotlin/moe/giga/discord/SettingsManager.kt
package moe.giga.discord

import io.github.cdimascio.dotenv.dotenv
import org.pmw.tinylog.Logger

class SettingsManager private constructor() {
    var settings: Settings
        private set

    private val dotenv = dotenv {
        directory = "./"
        ignoreIfMalformed = true
        ignoreIfMissing = true
    }

    private val envSettings = Settings(
            dotenv["DISCORD_TOKEN"],
            dotenv["DISCORD_CLIENT_ID"],
            dotenv["DISCORD_OWNER"]
    )

    init {
        Logger.info("Loading settings from environment.")
        this.settings = envSettings
    }

    private object Holder {
        val INSTANCE = SettingsManager()
    }

    companion object {
        val instance: SettingsManager by lazy { Holder.INSTANCE }
    }
}
